pkg/streamingpromql/types/ring_buffer_test.go

```diff
@@ -6,113 +6,174 @@ import (
 	"math"
 	"testing"
 
+	"github.com/prometheus/prometheus/model/histogram"
 	"github.com/prometheus/prometheus/promql"
 	"github.com/stretchr/testify/require"
 )
 
+// We want to ensure fpoint+hpoint ring buffers are tested consistently,
+// and we don't care about performance here so we can use an interface+generics.
+type ringBuffer[T any] interface {
+	DiscardPointsBefore(t int64)
+	Append(p T) error
+	ForEach(f func(p T))
+	UnsafePoints(maxT int64) (head []T, tail []T)
+	CopyPoints(maxT int64) ([]T, error)
+	LastAtOrBefore(maxT int64) (T, bool)
+	First() T
+	Reset()
+	GetPoints() []T
+	GetFirstIndex() int
+	GetTimestamp(point T) int64
+}
+
 func TestRingBuffer(t *testing.T) {
-	buf := NewRingBuffer(poolForRingBufferTesting{})
+	t.Run("test fpoint ring buffer", func(t *testing.T) {
```

Comment:
[nit] (same applies for other cases in this file)

```suggestion
	t.Run("test FPoint ring buffer", func(t *testing.T) {
```